druid/src/widget/textbox.rs

```diff
@@ -32,103 +29,54 @@ use crate::piet::{
 use crate::theme;
 use crate::widget::Align;
 
+use crate::text::{movement, offset_for_delete_backwards, EditableText, Movement, Selection};
+
 const BORDER_WIDTH: f64 = 1.;
 const PADDING_TOP: f64 = 5.;
 const PADDING_LEFT: f64 = 4.;
 
 // we send ourselves this when we want to reset blink, which must be done in event.
 const RESET_BLINK: Selector = Selector::new("druid-builtin.reset-textbox-blink");
 
-#[derive(Debug, Clone, Copy)]
-pub struct Selection {
-    /// The inactive edge of a selection, as a byte offset. When
-    /// equal to end, the selection range acts as a caret.
-    pub start: usize,
-
-    /// The active edge of a selection, as a byte offset.
-    pub end: usize,
-}
-
-impl Selection {
-    /// Create a selection that begins at start and goes to end.
-    /// Like dragging a mouse from start to end.
-    pub fn new(start: usize, end: usize) -> Self {
-        Selection { start, end }
-    }
-
-    /// Create a caret, which is just a selection with the same and start and end.
-    pub fn caret(pos: usize) -> Self {
-        Selection {
-            start: pos,
-            end: pos,
-        }
-    }
-
-    /// If start == end, it's a caret
-    pub fn is_caret(self) -> bool {
-        self.start == self.end
-    }
-
-    /// Return the smallest index (left, in left-to-right languages)
-    pub fn min(self) -> usize {
-        min(self.start, self.end)
-    }
-
-    /// Return the largest index (right, in left-to-right languages)
-    pub fn max(self) -> usize {
-        max(self.start, self.end)
-    }
-
-    /// Return a range from smallest to largest index
-    pub fn range(self) -> Range<usize> {
-        self.min()..self.max()
-    }
-
-    /// Constrain selection to be not greater than input string
-    pub fn constrain_to(mut self, s: &str) -> Self {
-        let s_len = s.len();
-        self.start = min(self.start, s_len);
-        self.end = min(self.end, s_len);
-        self
-    }
-}
-
 /// A widget that allows user text input.
 #[derive(Debug, Clone)]
-pub struct TextBox {
+pub struct TextBox<E: EditableText> {
     placeholder: String,
     width: f64,
     hscroll_offset: f64,
     selection: Selection,
     cursor_timer: TimerToken,
     cursor_on: bool,
+    phantom: std::marker::PhantomData<E>,
 }
 
-impl TextBox {
+impl<E: 'static + EditableText + Data + std::string::ToString> TextBox<E> {
```

Comment:
I think since we're somewhat compromising on the genericness of `EditableText`, we can get away with just keeping this as `TextBox<String>` for now.